ros2_foxglove_bridge/src/ros2_foxglove_bridge.cpp

```diff
@@ -814,6 +820,35 @@ void FoxgloveBridge::serviceRequest(const foxglove::ServiceRequest& request,
   client->async_send_request(reqMessage, responseReceivedCallback);
 }
 
+void FoxgloveBridge::fetchAsset(const std::string& uri, uint32_t requestId,
+                                ConnectionHandle clientHandle) {
+  foxglove::FetchAssetResponse response;
+  response.requestId = requestId;
+
+  try {
+    // We reject URIs that are not on the allowlist or that contain two consecutive dots. The latter
+    // can be utilized to construct URIs for retrieving confidential files that should not be
+    // accessible over the WebSocket connection. Example:
+    // `package://<pkg_name>/../../../secret.txt`. This is an extra security measure and should not
+    // be necessary if the allowlist is strict enough.
+    if (uri.find("..") != std::string::npos || !isWhitelisted(uri, _assetUriAllowlistPatterns)) {
+      throw std::runtime_error("Asset URI not allowed: " + uri);
+    }
+
+    const resource_retriever::MemoryResource memoryResource = _resource_retriever.get(uri);
```

Comment:
so this will block, even for HTTP fetches? Is there any built-in way to do this in a non-blocking fashion or should we consider doing it in a thread (if the API is thread safe)?

Reply:
You are right, this will block and prevent any other client operation from being handled since we use [only a single thread for handling client requests](https://github.com/foxglove/ros-foxglove-bridge/blob/9a206a06a2e8fd1564f066840a2b7e86d2cf74b7/foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp#L239-L240). I guess that this wouldn't be much of a problem for file:// or package:// URIs as these will succeed/fail fairly quickly, but for http it might indeed be problematic. I will see what I can do